AsyncDisplayKit/Details/ASCollectionViewFlowLayoutInspector.m

```diff
@@ -8,69 +8,172 @@
+  CGSize maxSize = collectionView.bounds.size;
+  if (ASScrollDirectionContainsHorizontalDirection(collectionView.scrollableDirections)) {
+    maxSize.width = FLT_MAX;
+  } else {
+    maxSize.height = FLT_MAX;
+  }
+  return ASSizeRangeMake(CGSizeZero, maxSize);
 }
 
-#pragma mark - Accessors
+#pragma mark - ASCollectionViewDefaultCustomLayoutInspector
 
-- (instancetype)initWithCollectionView:(ASCollectionView *)collectionView flowLayout:(UICollectionViewFlowLayout *)flowLayout;
+@implementation ASCollectionViewDefaultCustomLayoutInspector {
+  struct {
+    unsigned int implementsConstrainedSizeForNodeAtIndexPath:1;
+  } _dataSourceFlags;
+}
+
+#pragma mark Lifecycle
+
+- (instancetype)initWithCollectionView:(ASCollectionView *)collectionView
 {
   self = [super init];
+  if (self != nil) {
+    [self didChangeCollectionViewDataSource:collectionView.asyncDataSource];
+  }
+  return self;
+}
 
-  if (flowLayout == nil) {
-    ASDisplayNodeAssert(NO, @"Should never create a layout inspector without a layout");
+#pragma mark ASCollectionViewLayoutInspecting
+
+- (void)didChangeCollectionViewDataSource:(id<ASCollectionDataSource>)dataSource
+{
+  if (dataSource == nil) {
+    memset(&_dataSourceFlags, 0, sizeof(_dataSourceFlags));
+  } else {
+    _dataSourceFlags.implementsConstrainedSizeForNodeAtIndexPath = [dataSource respondsToSelector:@selector(collectionView:constrainedSizeForNodeAtIndexPath:)];
+  }
+}
+
+- (ASSizeRange)collectionView:(ASCollectionView *)collectionView constrainedSizeForNodeAtIndexPath:(NSIndexPath *)indexPath
+{
+  if (_dataSourceFlags.implementsConstrainedSizeForNodeAtIndexPath) {
+    return [collectionView.asyncDataSource collectionView:collectionView constrainedSizeForNodeAtIndexPath:indexPath];
   }
+  
+  return ASDefaultConstrainedSizeForNodeForCollectionView(collectionView);
+}
+
+- (ASSizeRange)collectionView:(ASCollectionView *)collectionView constrainedSizeForSupplementaryNodeOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath
+{
+  ASDisplayNodeAssert(NO, @"To support supplementary nodes in ASCollectionView, it must have a layoutInspector for layout inspection. (See ASCollectionViewFlowLayoutInspector for an example.)");
+  return ASSizeRangeMake(CGSizeZero, CGSizeZero);
+}
 
+- (NSUInteger)collectionView:(ASCollectionView *)collectionView numberOfSectionsForSupplementaryNodeOfKind:(NSString *)kind
+{
+  ASDisplayNodeAssert(NO, @"To support supplementary nodes in ASCollectionView, it must have a layoutInspector for layout inspection. (See ASCollectionViewFlowLayoutInspector for an example.)");
+  return 0;
+}
+
+- (NSUInteger)collectionView:(ASCollectionView *)collectionView supplementaryNodesOfKind:(NSString *)kind inSection:(NSUInteger)section
+{
+  ASDisplayNodeAssert(NO, @"To support supplementary nodes in ASCollectionView, it must have a layoutInspector for layout inspection. (See ASCollectionViewFlowLayoutInspector for an example.)");
```

Comment:
I know I should ignore the comment, but just to add my comment so we are on the same page. If ASDK users do not have any supplementary nodes in a custom layout these methods should never be called. In case they want to have supplementary nodes they have to define it's own layout inspector and implement the functions. They could actually also subclass this base layout inspector and just overwrite the supplementary nodes methods and get the automatic constrained size calculation for free.